lib/utils/authentication.js

```diff
@@ -44,23 +45,114 @@ module.exports.register = function(server, options, next) {
+                    algorithm: config.algorithm,
+                    issueTime: + new Date(),
+                    expireTime: + new Date() + config.hawkKeyLifespan,
+                    studyRoles: roleList
+                };
+
+                return credentials;
+            });
     };
 
     /**
+     * Retrieves the current roles for a user, revokes old roles, and
+     * adds the new roles
+     * @param {string} username - the username to retrieve roles
+     * @return {object} roles - the current roles
+     */
+    function refreshCurrentStudyRoles(username) {
+        const encryptedUsername = authUtils.encryptUsername(username);
+
+        const UserRole = server.plugins.bookshelf.model('UserStudyRole');
+    
+        return new UserRole()
+            .where({username: encryptedUsername})
+            .fetchAll({withRelated: 'role'})
+            .then((roleCollection) => {
+                return revokeRoles(username)
+                .then(() => {
+                    return addRolesFromCollection(username, roleCollection);
+                }).then(() => {
+                    return relations
+                        .studyAsync('Describe what %s can do', username)
+                        .then((studyRoles) => {
+                            return studyRoles;
+                        });
+                });
+            }).catch((err) => {
+                server.log(['error', 'refreshRoles'], err);
+                throw(err);
+            });
+    }
+    
+    /**
+     * helper function to revoke all roles from Relations
+     * @param {string} username - the username to retrieve roles
+     * @return {object} roles - the current roles
+     */
+    function revokeRoles(username) {
+        let revokes = [];
+        return relations.studyAsync('Describe what %s can do', username)
+            .then((roles) => {
+                _.each(roles, (value, key) => {
+                    _.each(value, (subValue) => {
+                        revokes.push(
+                            relations.studyAsync(
+                                '%s is not the %s of %s',
+                                 username,
+                                 subValue,
+                                 key
+                        ));
+                    });
+                });
+    
+                return Bluebird.all(revokes);
+            }).catch((err) => {
+                server.log(['error', 'revokeRoles'], err);
+                throw(err);
+            });
+    
+    }
+    
+    /**
+     * helper function to add all roles from a Bookshelf collection
+     * of database roles to Relations
+     * @param {string} username - the username to retrieve roles
+     * @param {collection} roleCollection - the Bookshelf collection of roles
+     * @return {object} roles - the current roles
+     */
+    function addRolesFromCollection(username, roleCollection) {
+        let roleAdditions = [];
+        roleCollection.each((value) => {
```

Comment:
This is a good place for a `map`, rather than pushing to `roleAdditions` within the `each`.

It's all semantics of course, but it's a good place to practice using maps if you want.